src/game/common/sparsematchfinder.h

```diff
@@ -17,18 +17,97 @@
 #include "always.h"
 #include <map>
 
-template<typename K, typename T> class SparseMatchFinder
+template<typename T, typename K> class SparseMatchFinder
 {
     class MapHelper
     {
     public:
         bool operator()(const K &left, const K &right) const { return left == right; }
     };
 
+    static int Count_Condition_Intersection(const K &key1, const K &key2) { return key1.Count_Intersection(key2); }
+    static int Count_Condition_Inverse_Intersection(const K &key1, const K &key2)
+    {
+        return key1.Count_Inverse_Intersection(key2);
+    }
+
 public:
-    // TODO implement more wrappers as they are determined required.
-    T &operator[](const K &key) { return m_bestMatches[key]; }
+    const T *Find_Best_Info(std::vector<T> const &vector, K const &key) const
+    {
+        auto iter = m_bestMatches.find(key);
+        const T *ret = nullptr;
+
+        if (iter != m_bestMatches.end()) {
+            ret = (*iter).second;
+        }
+
+        if (ret) {
+            return ret;
+        }
+
+        ret = Find_Best_Info_Slow(vector, key);
+        captainslog_dbgassert(ret, "no suitable match for criteria was found!\n");
+
+        if (ret) {
+            const_cast<std::map<K const, const T *, MapHelper> &>(m_bestMatches)[key] = ret;
```

Comment:
I have no clue what the original code did here but I know Find_Best_Info is const and I know m_bestMatches isn't volitile and this seems to work.

Reply:
Hmm. If `m_bestMatches` is not `mutable`, then how about removing the function const-ness?

Reply:
Because places that use it need the function to be const (and the things that use them and so on).